File: server/src/main/java/com/uiptv/service/BookmarkService.java
package com.uiptv.service;

import com.uiptv.db.BookmarkDb;
import com.uiptv.model.Bookmark;
import com.uiptv.model.BookmarkCategory;
import com.uiptv.util.AppLog;
import com.uiptv.util.ServerUtils;

import java.util.*;
import java.util.concurrent.CopyOnWriteArraySet;
import java.util.concurrent.atomic.AtomicLong;

public class BookmarkService {
    private final AtomicLong changeRevision = new AtomicLong(1);
    private final Set<BookmarkChangeListener> changeListeners = new CopyOnWriteArraySet<>();
    private volatile long lastUpdatedEpochMs = System.currentTimeMillis();

    private BookmarkService() {
    }

    public static BookmarkService getInstance() {
        return SingletonHelper.INSTANCE;
    }

    public boolean isChannelBookmarked(Bookmark bookmark) {
        return (BookmarkDb.get().getBookmarkById(bookmark) != null);
    }

    public Bookmark getBookmark(String dbId) {
        return (BookmarkDb.get().getById(dbId));
    }

    public Bookmark getBookmark(Bookmark bookmark) {
        return BookmarkDb.get().getBookmarkById(bookmark);
    }

    public void toggleBookmark(Bookmark bookmark) {
        Bookmark dbBookmark = BookmarkDb.get().getBookmarkById(bookmark);
        if (dbBookmark != null) {
            remove(dbBookmark.getDbId());
        } else {
            save(bookmark);
        }
    }

    public void save(Bookmark bookmark) {
        boolean created = BookmarkDb.get().save(bookmark);
        if (created) {
            int nextDisplayOrder = BookmarkDb.get().getNextDisplayOrder();
            BookmarkDb.get().saveBookmarkOrder(bookmark.getDbId(), nextDisplayOrder);
        }
        touchChange();
    }

    public List<Bookmark> read() {
        return BookmarkDb.get().getBookmarks();
    }

    public List<Bookmark> getBookmarksByCategory(String categoryId) {
        return BookmarkDb.get().getBookmarksByCategory(categoryId);
    }

    public void remove(String id) {
        try {
            BookmarkDb.get().delete(id);
            touchChange();
        } catch (Exception _) {
            AppLog.addErrorLog(BookmarkService.class, "Error while removing the bookmark");
        }
    }

    public String readToJson() {
        List<Bookmark> bookmarks = new ArrayList<>(BookmarkDb.get().getBookmarks());
        List<BookmarkResolver.ResolvedBookmark> resolved = new BookmarkResolver().resolveBookmarks(bookmarks);
        return ServerUtils.objectToJson(resolved.stream().map(BookmarkResolver.ResolvedBookmark::getBookmark).toList());
    }

    public String readToJson(int offset, int limit) {
        if (limit <= 0) {
            return readToJson();
        }
        int safeOffset = Math.max(0, offset);
        int safeLimit = Math.max(0, limit);
        List<Bookmark> bookmarks = new ArrayList<>(BookmarkDb.get().getBookmarksPage(safeOffset, safeLimit));
        List<BookmarkResolver.ResolvedBookmark> resolved = new BookmarkResolver().resolveBookmarks(bookmarks);
        return ServerUtils.objectToJson(resolved.stream().map(BookmarkResolver.ResolvedBookmark::getBookmark).toList());
    }

    // Category operations
    public List<BookmarkCategory> getAllCategories() {
        return BookmarkDb.get().getAllCategories();
    }

    public void addCategory(BookmarkCategory category) {
        BookmarkDb.get().saveCategory(category);
        touchChange();
    }

    public void removeCategory(BookmarkCategory category) {
        BookmarkDb.get().deleteCategory(category);
        touchChange();
    }

    // Order operations
    public void saveBookmarkOrder(List<String> orderedBookmarkDbIds) {
        Map<String, Integer> bookmarkOrders = new LinkedHashMap<>();
        for (int i = 0; i < orderedBookmarkDbIds.size(); i++) {
            bookmarkOrders.put(orderedBookmarkDbIds.get(i), i + 1);
        }
        saveBookmarkOrders(bookmarkOrders);
    }

    public void saveBookmarkOrders(Map<String, Integer> bookmarkOrders) {
        BookmarkDb.get().updateBookmarkOrders(bookmarkOrders);
        touchChange();
    }

    public long getChangeRevision() {
        return changeRevision.get();
    }

    public long getLastUpdatedEpochMs() {
        return lastUpdatedEpochMs;
    }

    public void addChangeListener(BookmarkChangeListener listener) {
        if (listener != null) {
            changeListeners.add(listener);
        }
    }

    public void removeChangeListener(BookmarkChangeListener listener) {
        if (listener != null) {
            changeListeners.remove(listener);
        }
    }

    private void touchChange() {
        lastUpdatedEpochMs = System.currentTimeMillis();
        long revision = changeRevision.incrementAndGet();
        for (BookmarkChangeListener listener : changeListeners) {
            try {
                listener.onBookmarksChanged(revision, lastUpdatedEpochMs);
            } catch (Exception _) {
                // Listener failures must never break bookmark mutation flow.
            }
        }
    }

    private static class SingletonHelper {
        private static final BookmarkService INSTANCE = new BookmarkService();
    }
}

File: server/src/main/kotlin/com/uiptv/service/BookmarkService.kt
package com.uiptv.service

import com.uiptv.db.BookmarkDb
import com.uiptv.model.Bookmark
import com.uiptv.model.BookmarkCategory
import com.uiptv.util.AppLog
import com.uiptv.util.ServerUtils
import java.util.ArrayList
import java.util.LinkedHashMap
import java.util.concurrent.CopyOnWriteArraySet
import java.util.concurrent.atomic.AtomicLong

object BookmarkService {
    private val changeRevision = AtomicLong(1)
    private val changeListeners = CopyOnWriteArraySet<BookmarkChangeListener>()
    @Volatile
    private var lastUpdatedEpochMs = System.currentTimeMillis()

    @JvmStatic
    fun getInstance(): BookmarkService = this
    fun isChannelBookmarked(bookmark: Bookmark): Boolean = BookmarkDb.get().getBookmarkById(bookmark) != null
    fun getBookmark(dbId: String?): Bookmark? = BookmarkDb.get().getById(dbId.orEmpty())
    fun getBookmark(bookmark: Bookmark): Bookmark? = BookmarkDb.get().getBookmarkById(bookmark)
    fun toggleBookmark(bookmark: Bookmark) {
        val dbBookmark = BookmarkDb.get().getBookmarkById(bookmark)
        if (dbBookmark != null) {
            remove(dbBookmark.dbId)
        } else {
            save(bookmark)
        }
    }
    fun save(bookmark: Bookmark) {
        val created = BookmarkDb.get().save(bookmark)
        if (created) {
            BookmarkDb.get().saveBookmarkOrder(bookmark.dbId.orEmpty(), BookmarkDb.get().getNextDisplayOrder())
        }
        touchChange()
    }
    fun read(): List<Bookmark> = BookmarkDb.get().getBookmarks()
    fun read(offset: Int, limit: Int): List<Bookmark> {
        if (limit <= 0) {
            return read()
        }
        val safeOffset = maxOf(0, offset)
        val safeLimit = maxOf(0, limit)
        return BookmarkDb.get().getBookmarksPage(safeOffset, safeLimit)
    }
    fun getBookmarksByCategory(categoryId: String?): List<Bookmark> = BookmarkDb.get().getBookmarksByCategory(categoryId)
    fun remove(id: String?) {
        try {
            BookmarkDb.get().delete(id.orEmpty())
            touchChange()
        } catch (_: Exception) {
            AppLog.addErrorLog(BookmarkService::class.java, "Error while removing the bookmark")
        }
    }
    fun removeByAccountName(accountName: String?) {
        try {
            BookmarkDb.get().deleteByAccountName(accountName.orEmpty())
            touchChange()
        } catch (_: Exception) {
            AppLog.addErrorLog(BookmarkService::class.java, "Error while removing bookmarks for account")
        }
    }
    fun readToJson(): String {
        val bookmarks = ArrayList(BookmarkDb.get().getBookmarks())
        val resolved = BookmarkResolver().resolveBookmarks(bookmarks)
        return ServerUtils.objectToJson(resolved.map { it.bookmark })
    }
    fun readToJson(offset: Int, limit: Int): String {
        if (limit <= 0) {
            return readToJson()
        }
        val safeOffset = maxOf(0, offset)
        val safeLimit = maxOf(0, limit)
        val bookmarks = ArrayList(BookmarkDb.get().getBookmarksPage(safeOffset, safeLimit))
        val resolved = BookmarkResolver().resolveBookmarks(bookmarks)
        return ServerUtils.objectToJson(resolved.map { it.bookmark })
    }
    fun getAllCategories(): List<BookmarkCategory> = BookmarkDb.get().getAllCategories()
    fun addCategory(category: BookmarkCategory) {
        BookmarkDb.get().saveCategory(category)
        touchChange()
    }
    fun removeCategory(category: BookmarkCategory) {
        BookmarkDb.get().deleteCategory(category)
        touchChange()
    }
    fun notifyBookmarksChanged() {
        touchChange()
    }
    fun saveBookmarkOrder(orderedBookmarkDbIds: List<String>) {
        val bookmarkOrders = LinkedHashMap<String, Int>()
        orderedBookmarkDbIds.forEachIndexed { index, dbId -> bookmarkOrders[dbId] = index + 1 }
        saveBookmarkOrders(bookmarkOrders)
    }
    fun saveBookmarkOrders(bookmarkOrders: Map<String, Int>) {
        BookmarkDb.get().updateBookmarkOrders(bookmarkOrders)
        touchChange()
    }
    fun getChangeRevision(): Long = changeRevision.get()
    fun getLastUpdatedEpochMs(): Long = lastUpdatedEpochMs
    fun addChangeListener(listener: BookmarkChangeListener?) {
        if (listener != null) {
            changeListeners.add(listener)
        }
    }
    fun removeChangeListener(listener: BookmarkChangeListener?) {
        if (listener != null) {
            changeListeners.remove(listener)
        }
    }

    private fun touchChange() {
        lastUpdatedEpochMs = System.currentTimeMillis()
        val revision = changeRevision.incrementAndGet()
        changeListeners.forEach { listener ->
            try {
                listener.onBookmarksChanged(revision, lastUpdatedEpochMs)
            } catch (_: Exception) {
                // Listener failures must never break bookmark mutation flow.
            }
        }
    }
}
